lib/grizzly/zwave_firmware.ex

```diff
@@ -0,0 +1,155 @@
+
+      {:error, :current_version_not_found} ->
+        Logger.warn("[Grizzly] Failed to get the hub's current firmware version")
+
+      {:error, :firmware_upload_failed} ->
+        Logger.warn("[Grizzly] Attempted to upload new firmware file but failed")
+    end
+
+    :ok
+  end
+
+  defp check_firmware_dir(%{zwave_firmware_dir: nil}) do
+    {:error, :no_firmware_dir}
+  end
+
+  defp check_firmware_dir(options) do
+    Logger.info("[Grizzly] Checking for firmware dir #{inspect(options.zwave_firmware_dir)}")
+
+    if File.exists?(options.zwave_firmware_dir) do
+      :ok
+    else
+      {:error, :firmware_dir_does_not_exist}
+    end
+  end
+
+  defp check_zwave_programmer_exists(options) do
+    if File.exists?(options.zw_programmer_path) do
+      :ok
+    else
+      {:error, :zw_programmer_does_not_exist}
+    end
+  end
+
+  defp get_firmware_file(options) do
+    dir = options.zwave_firmware_dir
+
+    case File.ls!(dir) do
+      [file] ->
+        {:ok, Path.join(dir, file)}
+
+      _ ->
+        {:error, :no_zwave_firmware_file}
+    end
+  end
+
+  defp maybe_upload_zwave_firmware_update(gbl_file, options) do
+    with {:ok, gbl_version} <- gbl_version(gbl_file),
+         {:ok, current_version} <- current_version(options) do
+      case validate_upgrade(current_version, gbl_version) do
+        :ok ->
+          upload_zwave_firmware_update(gbl_file, options)
+
+        {:error, :not_an_upgrade} ->
+          Logger.info("[Grizzly] The hub's zwave firmware is already up-to-date")
+          :ok
+      end
+    end
+  end
+
+  # .../ZW_SerialAPI_Controller_7.16.3_297_ZGM130S_REGION_US_LR.gbl
+  defp gbl_version(gbl_file) do
+    basename = Path.basename(gbl_file)
+    regex = ~r/\w+_(?<major>\d+)\.(?<minor>\d+)\.(?<patch>\d+).*\.gbl/
+
+    Regex.named_captures(regex, basename)
+    |> parse_version_capture()
+  end
+
+  defp current_version(options) do
+    with {answer, 0} <- System.cmd(options.zw_programmer_path, ["-s", options.serial_port, "-t"]),
+         {:ok, current_version} <- extract_current_version(answer) do
+      {:ok, current_version}
+    else
+      _other ->
+        {:error, :current_version_not_found}
+    end
+  end
+
+  # "Using serial device /dev/ttyS4\nConnected to Serial device: OK\nSerial version: 9, Chip type: 7, Chip version: 0, SDK: 7.15.02, ..."
+  defp extract_current_version(answer) do
+    Logger.info("[Grizzly] Extracting current version from #{inspect(answer)}")
+    regex = ~r/.*SDK:\s*(?<major>\d+)\.(?<minor>\d+)\.(?<patch>\d+)/m
+
+    Regex.named_captures(regex, answer)
+    |> parse_version_capture()
+  end
+
+  # Can not parse using Version.parse/1 because it fails on "7.15.02" etc.
+  defp parse_version_capture(%{"major" => major_s, "minor" => minor_s, "patch" => patch_s}) do
+    {major, ""} = Integer.parse(major_s)
+    {minor, ""} = Integer.parse(minor_s)
+    {patch, ""} = Integer.parse(patch_s)
+    {:ok, %Version{major: major, minor: minor, patch: patch}}
```

Comment:
The `Version` struct isn't friendly to manual creation. The only safe way is to call `Version.parse/1` unfortunately.

Reply:
@jfcloutier it looks like we might want to do this:

```elixir
case Version.parse("#{major_s}.#{minor_s}.#{patch_s}") do
  :error -> {:error, :invalid}
  ok_version -> ok_version
end
```

Another small update would be to move away from named captures and just do a string capture on `\d+.\d+.\d+`


If you want to be able to unit test some of this functionality could add another module just for version parsing or you can change the associated `defp` functions to be `def` functions and add `@doc false` above them.

Reply:
Unfortunately zw_commander currently returns the version string `7.15.02` which Version considers invalid. Which is why I can't use `Version.parse`.

`iex(1)> Version.parse("7.15.02")
:error
`

Reply:
Yeah, I see. Elixir's Version module follows semver which does not all for leading zeros. That is a bummer they report non semver.

Reply:
Can you add a comment for this code that explains that we are not parsing the Z-Wave firmware version using the `Version` module because Z-Wave does not report the version in a way that is compatible with semver leading to `Version.parse/1` to return `:error`?